lib/bitcoin/versions/latest/NormalizedFeeCalculator.ts

```diff
@@ -21,19 +26,46 @@ export default class NormalizedFeeCalculator implements IFeeCalculator {
     console.log(`Initializing normalized fee calculator.`);
   }
 
+  /**
+   * This adds normalized fee to the block as it would calculate normalziedFee, but uses a cache to remeber previously seen blocks.
+   * Which reduces calls to the metadata store.
+   */
+  public async addNormalizedFeeToBlock (blockMetadata: BlockMetadataWithoutNormalizedFee): Promise<BlockMetadata> {
+    const shouldCalculate = this.shouldCalculateNormalizedFee(blockMetadata.height);
+    if (!shouldCalculate.shouldCalculate) {
+      return Object.assign({ normalizedFee: shouldCalculate.fee! }, blockMetadata);
+    }
+    // the cache won't work if the block is not expected, refetch the blocks and store in cache
+    if (this.expectedBlockHeight !== blockMetadata.height) {
+      this.blockMetadataCache = await this.getLookBackBlocks(blockMetadata.height);
+      this.expectedBlockHeight = blockMetadata.height;
+    }
+    const newFee = this.calculateNormalizedFee(this.blockMetadataCache);
+    const newBlockWithFee = Object.assign({ normalizedFee: newFee }, blockMetadata);
+    this.blockMetadataCache.push(newBlockWithFee);
+    this.blockMetadataCache.shift();
+    this.expectedBlockHeight++;
+    return newBlockWithFee;
+  }
+
   public async getNormalizedFee (block: number): Promise<number> {
-    // DB call optimization
-    // https://github.com/decentralized-identity/sidetree/issues/936
+    const shouldCalculate = this.shouldCalculateNormalizedFee(block);
+    if (!shouldCalculate.shouldCalculate) {
+      return shouldCalculate.fee!;
+    }
+    const blocksToAverage = await this.getLookBackBlocks(block);
+    return this.calculateNormalizedFee(blocksToAverage);
+  }
+
+  private shouldCalculateNormalizedFee (block: number): { shouldCalculate: boolean, fee: number | undefined } {
```

Comment:
Please give me a better name (cry)

Reply:
LOL. I'd suggest eliminating this method, and just implement the logic in both callers. I double checked, you don't need to handle pre-genesis case in both cases (`getNormalizedFee()` in `BitcoinProcessor` already does the same check and actually throws error which I think is the better behavior than returning 0), so code there would be more or less the same even without this method.

Reply:
Oh, true. Ok.